### PR TITLE
Feature/20459 sub steps refactor

### PR DESCRIPTION
**Pull request by burczu**
### Details

In this PR I've refactored some parts of the code related to sub-steps forms that we recently use more extensively in the App. These improvements was discussed with @koko57 and @MrMuzyk - it will be useful in the Global Reimbursement project where we create more sub-steps forms. The PR introduces few re-usable common form components like `FullNameStep`, `DateOfBirthStep`, `AddressStep` or `SingleFieldStep` (that can be used for e.g. phone numbers or social security numbers) ones and uses them wherever applicable. It also makes the InteractiveStepWrapper component more re-usable.

### Fixed Issues

$ https://github.com/Expensify/App/issues/50893
PROPOSAL: n/a

### Tests

Enable payments:

1. Log in to the App
2. Go to Settings -> Wallet
3. Add bank account
4. Click "Enable wallet"
5. Test all the "Personal info" steps linke full name, address, date of birth

Reimbursement account - personal info:

1. Log in to the App
2. Create new workspace
3. Add new member
4. Go to More features section and enable Expensify Card option
5. Go to Expensify Card section that just appeared
6. Tap Issue new card
7. Go through the process of connecting to the bank account etc. (use testing data described in this Slack thread: https://callstack-hq.slack.com/archives/C01GTK53T8Q/p1711049879962529)
8. At some point of the process you'll need to provide requestor information like first and last name, date of birth etc. - test if it works correctly

Reimbursement account - beneficial owner info:

1. Repeat steps 1-7 described above
2. On the later steps of the process, you'll be asked: "Do you own 25% or more of Alberta Bobbeth Charleson?" - select "No"
3. On the next step: "Do any individuals own 25% or more of Alberta Bobbeth Charleson?" - select "Yes"
4. You'll land on the next personal info multiple steps form - test it if it works correctly

Missing personal info for issuing the card:

1. Log in to the App as User A
2. Go to the Workspace that you are admin of
3. Invite completely new user (User B)
4. Turn on Expensify Card for this workspace if not done yet
5. Set it up
6. Tap "Issue new card" and issue the card for User B
7. Log in to the App as User B
8. Go to the Workspace chat - you should see the "Add shipping details" button - tap it
9. You'll see the multiple step form asking you to fill your personal details - test it

- [x] Verify that no errors appear in the JS console

### Offline tests

Same as above

### QA Steps

Same as above

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

Enable payments:

https://github.com/user-attachments/assets/044f2626-85d8-4413-8f47-23ad15777309

Reimbursement account - personal info:

https://github.com/user-attachments/assets/00dfb24c-1195-42eb-bd4f-4bf873673405

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/0efbdb9a-3cdc-43cd-a773-3a87638fa3ce

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/f70678b5-a368-42aa-af3d-96e0c07f018a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

Enable payments:

https://github.com/user-attachments/assets/dc732b4b-5df7-41a1-b545-5ce2b9f53845

Reimbursement account - personal info:

https://github.com/user-attachments/assets/eb405376-c2b2-44c5-81ec-ccfd09d7015e

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/b1250440-2a30-47bc-b267-66abba4eedd5

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/20753138-2ba9-4da4-a6e0-7a3f088a7728

</details>

<details>
<summary>iOS: Native</summary>

Enable payments:

https://github.com/user-attachments/assets/db2721ca-d85b-4dfc-b617-e849448e7072

Reimbursement account - personal info:

https://github.com/user-attachments/assets/65b3c078-5757-4a97-be9b-c4adf1504c2d

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/8c51a034-f6c5-4644-9d75-241a38dc7956

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/67a9711c-ca83-4c2a-a43b-e8c369ef3ee5

</details>

<details>
<summary>iOS: mWeb Safari</summary>

Enable payments:

https://github.com/user-attachments/assets/0b687827-f559-471b-b99a-cb3a5d885e9f

Reimbursement account - personal info:

https://github.com/user-attachments/assets/1372e8e1-dbc2-44dc-af9b-e20e0e108d49

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/a19e1d66-520c-49af-9c31-e26d69c7319c

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/e2e0d5e2-a81d-438f-a49b-2a89e6f7158a

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

Enable payments:

https://github.com/user-attachments/assets/e344bf3b-0078-42ec-a72c-4bec7d94cefe

Reimbursement account - personal info:

https://github.com/user-attachments/assets/b8cd0246-52d3-4c2a-bdf6-cf1ab75263d1

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/f4937652-df81-4078-8a63-82365c85102a

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/2e2c64b8-1120-4ea8-813b-e577d359b703

</details>

<details>
<summary>MacOS: Desktop</summary>

Enable payments:

https://github.com/user-attachments/assets/cc7662da-21ab-4a29-b1d2-6a0112564ea4

Reimbursement account - personal info:

https://github.com/user-attachments/assets/a3ac6bdb-70a7-4613-89db-83752a76629f

Reimbursement account - beneficial owner info:

https://github.com/user-attachments/assets/9b25e56b-59ac-4de8-91e1-5f1db61e907c

Missing personal info for issuing the card:

https://github.com/user-attachments/assets/37c0ede3-0f31-47e2-bf6d-7e6ff0f1dc33

</details>

### CODE REVIEW

**Comment by koko57** [comment]
at first glance looks nice! great job!

**Comment by hungvu193** [comment]
I'll review this one this weekend 😄

**Comment by hungvu193** [comment]
@burczu  Is it expected that we should navigate to `Connect Bank Account` page on pressing back?

https://github.com/user-attachments/assets/569f6239-be4e-4432-830e-b6b2f19dbca7

File: src/components/SubStepForms/ConfirmationStep.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        <SafeAreaConsumer>
+            {({safeAreaPaddingBottomStyle}) => (
+                <ScrollView
+                    style={styles.pt0}
+                    contentContainerStyle={[styles.flexGrow1, safeAreaPaddingBottomStyle]}
+                >
+                    <Text style={[styles.textHeadlineLineHeightXXL, styles.ph5, styles.mb3]}>{pageTitle}</Text>
+                    {summaryItems.map(({description, title, shouldShowRightIcon, onPress}) => (
+                        <MenuItemWithTopDescription
+                            description={description}
```
Comment (hungvu193) [inline src/components/SubStepForms/ConfirmationStep.tsx:58]: I think we should add a `key` props here.

**Review by hungvu193 — COMMENTED** [review]
Great works! I think we have [1 bug](https://github.com/Expensify/App/pull/51231#issuecomment-2440691636) left.

File: src/components/SubStepForms/ConfirmationStep.tsx (hunk 1 above)
Comment (burczu) [reply src/components/SubStepForms/ConfirmationStep.tsx:59]: good point, thanks - added missing `key` prop

**Comment by burczu** [comment]
> @burczu Is it expected that we should navigate to `Connect Bank Account` page on pressing back?

I think so. Also, I don't think I've touched this behaviour while working on refactor.

**Comment by burczu** [comment]
> Bug: In Personal Info, while entering invalid zipcode and press Submit, error is shown at address field.

Right, that was a bug - fixed now. Thanks

**Comment by hungvu193** [comment]
> > Bug: In Personal Info, while entering invalid zipcode and press Submit, error is shown at address field.
> 
> Right, that was a bug - fixed now. Thanks

Cool. Can you merge main? I'll verify in few miunutes

**Comment by burczu** [comment]
> Cool. Can you merge main? I'll verify in few miunutes

done

**Comment by hungvu193** [comment]
For some reason I can't seem to text on simulator :/, recording screenshots for all platforms will take more time than I expected

**Comment by hungvu193** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



https://github.com/user-attachments/assets/86e2afc7-e198-43b0-8efa-4e592c682940


</details>

<details>
<summary>iOS: Native</summary>




https://github.com/user-attachments/assets/ea02d10d-6d8b-4e4b-bb1c-9f1a2cb3a140



</details>

<details>
<summary>iOS: mWeb Safari</summary>



https://github.com/user-attachments/assets/6fe74de8-4e58-406e-b912-ca6b02ffcbf0


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



https://github.com/user-attachments/assets/2ae2c77c-c929-4645-90a0-3fb1d56fc356


https://github.com/user-attachments/assets/7086219b-be22-4055-af1d-1754e56c1314


https://github.com/user-attachments/assets/44415547-734b-464f-9a5d-3c4dbbc8b814


</details>

<details>
<summary>MacOS: Desktop</summary>



https://github.com/user-attachments/assets/f3f1d877-a3b1-4c28-a30b-b851e5249260

</details>

**Review by hungvu193 — APPROVED** [review]

**Comment by hungvu193** [comment]
@burczu  Sorry, I forgot the physical back button press on Android. Do you think we also need to handle it? Please take a look at the video below, if we press Back button, the screen will be existed instead of navigating to previous step.

https://github.com/user-attachments/assets/a5b1382e-578e-4947-b898-cea036ecc17e

**Comment by burczu** [comment]
> @burczu Sorry, I forgot the physical back button press on Android. Do you think we also need to handle it? Please take a look at the video below, if we press Back button, the screen will be existed instead of navigating to previous step.

Hmm... I think this ineed might be a bug, but I don't think it is related to this refactor - I did not touch this kind logic here.

**Comment by hungvu193** [comment]
I see. Then I think we're good to go here  👍

**Review by madmax330 — APPROVED** [review]

**Pull request merged**